### ISSUE TITLE
all wsl.exe commands hang

### ISSUE THREAD

**Issue by devsnek** [microsoft/WSL#4144]
**Please fill out the below information:**
* Your Windows build number:  (Type `ver` at a Windows Command Prompt)

Microsoft Windows [Version 10.0.18917.1000]

* What you're doing and what's happening:  (Copy&paste the full set of _specific_ command-line steps necessary to reproduce the behavior, and their output. Include screen shots if that helps demonstrate the problem.)

`$ wsl.exe --help`
`$ wsl.exe -d Debian`
etc...

Example with `--help` passed (screenshot taken 30s after running command)
![](https://gc.gy/bb19f80d-d0ac-4c1e-8e71-9df61489908f.png)

* What's wrong / what should be happening instead:

`wsl.exe` should not hang.

**Comment by onomatopellan** [comment]
Run `wsl.exe --shutdown` on an elevated cmd to make sure there is no VM running that can slowdown everything.

Things to try:
-what it shows on cmd when you type `where wsl` ?
-on powershell does `wsl --help` hangs too?
-show the output of `wsl -l -v` and `wsl echo "Hello"`. Did they hang too?

**Comment by devsnek** [comment]
@onomatopellan everything you suggest also hangs

**Comment by devsnek** [comment]
restarting fixed this, but obviously i don't know if it will happen again...

**Comment by onomatopellan** [comment]
I'm running 18917 inside a Hyper-V VM and it didn't happen to me yet.

Next time you should use Process Explorer to see what's happening during the hangs.

**Comment by mscraigloewen** [comment]
Seems related to issue #4047.

**Comment by kamil-mrzyglod** [comment]
It occasionally happens on my PC running 10.0.17763.740.

**Comment by rlipscombe** [comment]
Seeing this with 10.0.19013.1.

**Comment by arsdever** [comment]
Observed 10.0.19041.388.

- No response to any command. Just hangs
- `where wsl` gives nothing

**Comment by kthompson** [comment]
Also experiencing this with `10.0.18363.1049`

**Comment by 8549** [comment]
I'm also running into this problem on 10.0.20211, restarting the host machine (maybe a couple of times) usually does the trick.

**Comment by emilsrits** [comment]
Started happening recently, all `wsl.exe` related commands hang except for `wsl --help`.
OS version: 1909
OS build: 18363.1082

**Comment by aadibajpai** [comment]
same issue, except wsl --help is hanging as well

**Comment by nickb937** [comment]
I had this problem too.  Trying to install an image would get stuck in the Installing state, or converting an image from WSL1 would report the error "the operation could not be started because a required feature is not installed."

The suggestion of disabling the override of Control Flow Guard in this comment solved the problem: https://github.com/microsoft/WSL/issues/4626#issuecomment-553416411

**Comment by taseikyo** [comment]
same issue, all `wsl` commands (including `wsl --help`) hang
OS version: 1909
OS build: 18363.1198

**Comment by jsteenri** [comment]
same issue, all wsl commands (including wsl --help) hang for a long time and then return with:
```
PS C:\WINDOWS\system32> wsl --help
Server execution failed
PS C:\WINDOWS\system32>
```

OS: Windows 10
OS version: 2004
OS build: 19041.572

No updates to Windows in past eight days, yet this issue begins out of the blue...

**Comment by jsteenri** [comment]
Update: Disabling Hyper-V seems to have solved this issue.      _Or rebooting six or eight times; I guess I won't ever know._  
  
**Details**
[ServerFault issue](https://superuser.com/questions/1595011/windows-update-broke-windows-subsystem-for-linux-2?newreg=d5c1ccf23c0b4c8e84dd7dfd91efea92) mentioned disabling and re-enabling WSL, but I started by first using 'optionalfeatures.exe' from Command Prompt as described there, yet only unchecking entire "Hyper-V" section and rebooting [per this list of suggestions](https://github.com/microsoft/WSL/issues/6059#issuecomment-706327392)  

**Results**
Now, WSL2 distro is working, all files intact, and Docker Desktop (using WSL2 backend) is also happy.    Note, Hyper-V had been enabled, though largely unused, for all of 2020, while WSL2 usage has only been since August.   So maybe Hyper-V and WSL don't play nice together, or something from a recent update bungled that?

**Theory**
Something with Hyper-V (and maybe a recent update) was causing the LxssManager service to be dead upon Windows startup and completely unresponsive to any changes.   With Hyper-V now out-of-the-way, Lxssmanager once again appears healthy.

**Pro Tip**
Although I didn't need it, I felt better making the above change(s) after creating a backup of the WSL2 filesystem [as described in this comment](https://github.com/microsoft/WSL/issues/6059#issuecomment-705827714).  The one for Docker is similarly located at: C:\Users\<you>\AppData\Local\Docker\wsl\data

**Comment by bpulliam** [comment]
This is another long-running issue that encapsulates many separate issues with a similar symptom.  Since the original issue, we have fixed many things that could display that symptom.  I'm going to close this issue so that we can accurately track new issues and not lose anything.  If you see an issue with unstable Hyper-V or a hang in LxssManager, please file a new issue so we can gather logs and get it fixed.  Thanks!

**Comment by Neutrovertido** [comment]
Managed to fix the issue by disabling all WSL related features, rebooting, then enabling them again and rebooting once more.
Later I had to run `wsl --set-default-version 2` again.

**Comment by JohnCido** [comment]
> Managed to fix the issue by disabling all WSL related features, rebooting, then enabling them again and rebooting once more. Later I had to run `wsl --set-default-version 2` again.

Disabled Hyper-V plus WSL -> reboot -> enable them again -> reboot solved my issue!

**Comment by jordan69420** [comment]
I reproduced the issue, and it stops all commands from working, try not breaking your WSL dist.

**Comment by jordan69420** [comment]
also try restarting your PC

**Comment by tommai78101** [comment]
@JohnCido Can confirm, I did what you said, to fix my issue after upgrading from Windows 10 to Windows 11. Docker Desktop, Hyper-V, and WSL 2 were already installed on Windows 10 prior to the upgrade. I wondered if a change in OS version causes the WSL to fail?

**Comment by kedare** [comment]
I have this issue on fresh Windows 11 install (not upgraded from Windows 10), so it's probably not related to a previous upgrade.

**Comment by Mistic92** [comment]
I have the same issue now. WSL2 often just got stuck and even help commands does not work at all
![image](https://user-images.githubusercontent.com/2607098/162700379-4ed30bac-75de-436f-bcb0-f9c72f5b0f1b.png)

**Comment by XenoBino** [comment]
As of 23 __Apr 2022__, this issue persists. I have encountered this issue a couple of times since I started using WSL in late 2020.
I am using Windows 11 21H2 OS Build 22000.613, WSL Kernel 5.10.102.1. There are no updates to Windows nor WSL.
This issue is really frustrating as the only known solution is to restart.
I hope it gets series attention.

**Comment by inaroquet** [comment]
WSL worked on my machine on Friday. I shut down for the weekend and started up on Monday and I experienced the issue described in this thread.

WSL Ubuntu-20.04 hangs when try to start Ubuntu from the windows search menu
wsl --help: Does not hang
wsl -l -v: Hangs, interruptible
wsl --shutdown: Hangs, interruptible
Docker Desktop which is set to run on WSL backend hangs on launch

I shut down my pc and started it after 30 minutes but this WSL issue persisted. 
I restarted my pc and then the issue was resolved and WSL Ubuntu and Docker Desktop worked.

**Comment by andrew121410** [comment]
This is beyond annoying. Yeah everything is hanging all WSL commands. Restarting my computer still doesn't fix it.

**Comment by tommai78101** [comment]
@andrew121410 File a new issue. They are no longer tracking this issue.

**Comment by ismailhkose** [comment]
@[bpulliam](https://github.com/bpulliam)
experienced the hanging issue at Version 10.0.22000.739.
Every command starts with `wsl` hangs. After waiting for a long time, I occasionally saw `Server execution failed` error messages.
I restarted a few times but did not help.

[bpulliam](https://github.com/bpulliam), Please reach me out privately, via Teams, so I can provide the necessary telemetries for debugging the issue.

**Comment by Sollace** [comment]
> 

The above appears to be exactly what I needed to fix this! After doing the above to unblock it, the first thing I went to check was that I had properly updated wsl, so did that, and then checked what version I was running with `wsl -l -v`. Turns out I had accidentally set up a wsl 1 system yesterday, which may be the reason for it hanging now.

I'm updating it now to wsl 2, but for anyone else having the same problem, perhaps check that you're not _also_ running wsl 1, because that might be the issue. I had no hanging problems before yesterday, so a solution may just be to upgrade your install.

**Comment by JohnCido** [comment]
Also see this issue: #5845
Like @WinniX  mentioned, it could be related to a `LxssManager` problem
> Had the same issue. To fix it I first went to Device Manager and noticed that "LxssManager" is trying to stop, but it can't. To unblock it I did this. Run the following command to find the PID of the "LxssManager" process: `tasklist /svc /fi "imagename eq svchost.exe"` Then went to Task Manager and killed the svchost with this PID. After that I saw that "LxssManager" is unblocked and finally running. Now I can run `wsl` commands again 😅

**Comment by grandchild** [comment]
Even shorter output than what @WinniX posted:
```cmd
C:\> tasklist /nh /svc /fi "services eq lxssmanager"
svchost.exe            12345 LxssManager
```

Then I'd do
```cmd
taskkill /f /pid 12345
```
but unfortunately, that tells me "access denied" 🙁

But it works from the TaskManager/Details tab. So hey -- at least no reboot required.

**Comment by BennyBread** [comment]
Killing the LxssManager did the trick for me as well! 

#Team-NoReboot 😅

**Comment by i3v** [comment]
> but unfortunately, that tells me "access denied" even from an admin cmd 🙁

[Processhacker](https://processhacker.sourceforge.io/) just worked for me in this case.

**Comment by wqlu** [comment]
> > Managed to fix the issue by disabling all WSL related features, rebooting, then enabling them again and rebooting once more. Later I had to run `wsl --set-default-version 2` again.
> 
> Disabled Hyper-V plus WSL -> reboot -> enable them again -> reboot solved my issue!

this worked, and solved my problem.

**Comment by juniortarcisio** [comment]
> Had the same issue. To fix it I first went to Device Manager and noticed that "LxssManager" is trying to stop, but it can't. To unblock it I did this. Run the following command to find the PID of the "LxssManager" process: `tasklist /svc /fi "imagename eq svchost.exe"` Then went to Task Manager and killed the svchost with this PID. After that I saw that "LxssManager" is unblocked and finally running. Now I can run `wsl` commands again 😅

This worked for me, thanks.

**Comment by alkuzad** [comment]
method with killing svhost is good but I had to kill more than just LxssManager svhost, there also were wsl processes hanging out in there. So for my case:

Open elevated Process Explorer, switch to process tree
Stop everything related to WSL and LxssManager under svhost and explorer.exe trees
Restart-Service LxssManager

**Comment by canercetin-randomguy** [comment]
Error still persists. Everything hangs until I kill LxssManager.

like, what the hell. Is there anything stable in WSL? 
All of a sudden
-> WSL commands may hang
-> WSL may take %95 of the CPU forcing you to restart the computer
-> WSL may screw your entire storage and never shrinking back automatically
-> services like Docker in WSL may not work with VPN, and even after you close VPN, you have to set iptables with

`sudo update-alternatives --set iptables /usr/sbin/iptables-legacy
sudo update-alternatives --set ip6tables /usr/sbin/ip6tables-legacy
`

everything I listed is still going on.

https://github.com/microsoft/WSL/issues/4699

https://github.com/microsoft/WSL/issues/6982

https://github.com/microsoft/WSL/issues/4144

https://github.com/microsoft/WSL/issues/5068

buuuuut, hey! Cortana and other bloats are working perfectly fine and stable without errors, yay Windows!

**Comment by adityacryptota** [comment]
For me the thing that fixed it was ending the process wsl.exe and wslhost.exe. You should see VmmemWSL in the task manager. Click on that and go to details. You should now see wsl.exe and wslhost.exe. End all those processes.

**Comment by alkuzad** [comment]
I created gist that kill/restart everything related to Docker/WSL and restarts, for me, regardless of the source issue - this successfully unlocks docker - https://gist.github.com/alkuzad/438993ecfea601cd56536bc360e5f581

**Comment by FLAMESpl** [comment]
I have neiher Hyper-V enabled nor lxssmanager running yet my WSL is totally dead. Any tips for me?

**Comment by FLAMESpl** [comment]
`taskkill /f /im wslservice.exe` did the trick

**Comment by canercetin-randomguy** [comment]
> Had the same issue. To fix it I first went to Device Manager and noticed that "LxssManager" is trying to stop, but it can't. To unblock it I did this. Run the following command to find the PID of the "LxssManager" process: `tasklist /svc /fi "imagename eq svchost.exe"` Then went to Task Manager and killed the svchost with this PID. After that I saw that "LxssManager" is unblocked and finally running. Now I can run `wsl` commands again 😅

Issue still persists, and I am tired of doing this over and over again when everything hangs. Becomes annoying reaaaally quick.

and I was here last month again. this is the only reason I choose bare metal over WSL, this critical issue is here FOR ONE WHOLE YEAR.

and yeah this comment is for bumping the issue.

**Comment by ryandwalker** [comment]
Shutting down and then booting back up did not fix the issue.  But a "restart" immediately fixed the issue.  Apparently it is not always the same in Windows to shutdown and boot up again as it is to restart.  I tried several times to shutdown and then start and that didn't work, but one true "restart" of Windows fixed it.

**Comment by aldobarr** [comment]
> Shutting down and then booting back up did not fix the issue. But a "restart" immediately fixed the issue. Apparently it is not always the same in Windows to shutdown and boot up again as it is to restart. I tried several times to shutdown and then start and that didn't work, but one true "restart" of Windows fixed it.

You need to untick "Turn on fast startup" in control panel -> hardware and sound -> power options -> system settings
in order for a shutdown to do what you expect. Otherwise you need to restart for windows to truly turn off and reset.

**Comment by natechoiniere** [comment]
> You need to untick "Turn on fast startup" in control panel -> hardware and sound -> power options -> system settings
in order for a shutdown to do what you expect. Otherwise you need to restart for windows to truly turn off and reset.

There is no such option.

**Comment by DAMACANER** [comment]
People still fly by to this issue and can replicate it, why is this closed?

**Comment by t-money-g** [comment]
This is still happening on 19044.3086  on Windows Enterprise.

**Comment by migounette** [comment]
@guillemdj's comment works for me

```powershell
C:\Windows\System32>taskkill /f /im wslservice.exe
SUCCESS: The process "wslservice.exe" with PID 6424 has been terminated.
```

During this operation, a crash occured... it may help WSL team to understand the issue
![image](https://github.com/microsoft/WSL/assets/1232978/097fb3aa-c92e-4a2c-a127-9c6c8cdb5b78)


```poweshell
C:\Windows\System32>wsl --status
Default Distribution: Ubuntu-20.04
Default Version: 2
```

You can also collect logs with - https://github.com/microsoft/WSL/blob/master/diagnostics/collect-wsl-logs.ps1

**Comment by domenicquirl** [comment]
Just encountered this on `10.0.22000.2295`. I got it working again going off of [this comment](https://github.com/microsoft/WSL/issues/4144#issuecomment-1185588635), but `LxssManager` wasn't even running when I first checked. So I had to manually start it (which didn't help yet), _then kill it_, and only did the `wsl` commands stop hanging.

**Comment by diamondzxd** [comment]
Happened just now, all wsl commands hanging except `wsl --help`

**Comment by Parideboy** [comment]
> Happened just now, all wsl commands hanging except `wsl --help`

Same here, can't use wsl anymore, no matter what I try

**Comment by diegocaprioli** [comment]
> Managed to fix the issue by disabling all WSL related features, rebooting, then enabling them again and rebooting once more. Later I had to run `wsl --set-default-version 2` again.

This worked for me. Thanks!

**Comment by NewEpoch2020** [comment]
I used the following command to shut down the LxssWanager service to solve the problem.
 `for /f "tokens=2" %a in ('tasklist /svc /fi "imagename eq svchost.exe" ^| findstr LxssManager') do pskill -t %a`

**Comment by syamsoul** [comment]
> Managed to fix the issue by disabling all WSL related features, rebooting, then enabling them again and rebooting once more. Later I had to run `wsl --set-default-version 2` again.

Thank you.... it worked for me..

1. wsl --set-default-version 2
2. reboot

**Comment by DanKaplanSES** [comment]
@grandchild 
> Even shorter output than what @WinniX posted:
> ...
> But it works from the TaskManager/Details tab. So hey -- at least no reboot required.

I'm on Windows 10 so that might be the difference, but when I tried to kill this in the task manager, it displays me a pop up that _forces_ me to restart:

![image](https://github.com/microsoft/WSL/assets/373675/1af01b6c-f8bd-41bb-aa70-b375d4a55de9)

I'm going to try it anyway; wish me luck.

**Comment by mback4** [comment]
I can really only replicate this problem by opening vscode from wsl/ubuntu and then forgetting to close the window when I let my computer go to sleep at the end of the workday. Coming back typically wsl has hung, won't respond to `wsl --shutdown` at all. 

Using `taskkill` to terminate the wslservice.exe fixes the problem for me.

**Comment by 1EDExg0ffyXfTEqdIUAYNZGnCeajIxMWd2vaQeP** [comment]
Killing `wslservices.exe` is not a good workaround, as it results in a bluescreen.

**Comment by caner-cetin** [comment]
> Killing `wslservices.exe` is not a good workaround, as it results in a bluescreen.

thought my gmail glitched wtf is that name lmao

**Comment by hicotton02** [comment]
I have the same issue. I have even uninstalled any distros, disabled all features, rebooted, enabled the features, rebooted, and tried to reinstall the distro. It hangs still. I am on Win 11 with latest updates. At this point, I am looking at reinstalling Windows....

**Comment by sedlund** [comment]
for me I had `rclone` configured as a service that mounts a remote network storage 
 as a drive.  and that storage was down. 

this caused all the wsl commands (`wsl -l` `wsl --shutdown`) to hang and WSL from starting.

after stopping the service wsl began working again.

**Comment by maxfelker** [comment]
Having the same issue with Window 10 Version 22H2 OS build 19045.4651. My issue was the my physical hard drive ran out of memory while I was using VS Code and building docker images. This comment tipped me off:

> `taskkill /f /im wslservice.exe` did the trick

I right clicked on my terminal and _Ran as Adminstrator_ then ran the command above. In the same terminal I then ran `wsl`. 

From there I was able to open WSL by either opening up the _Ubuntu_ app from the start bar or typing `wsl` into my terminal running at user level. That resolved my issue, did not get BSoD.

**Comment by HernandoR** [comment]
Constantly experience this Issue, regardless reboots.

**Comment by vegerot** [comment]
I am still getting this issue.  Every wsl command hangs, including commands to kill wsl

**Comment by karmamaster** [comment]
Nothing of above work, anyone have ways to work around?

**Comment by elide-b** [comment]
Give the WSL command enough time to time out, even up to 10 minutes, try disabling the features and restarting, and eventually you might be lucky enough to see a message like this:

```
wsl: WSL installation appears to be corrupted (Error code: Wsl/CallMsi/Install/REGDB_E_CLASSNOTREG).
Press any key to repair WSL, or CTRL-C to cancel.
This prompt will time out in 60 seconds.
Class not registered
Error code: Wsl/CallMsi/Install/REGDB_E_CLASSNOTREG
```

and pressing a key then actually updates and repairs the WSL installation

**Comment by syamsoul** [comment]
> > Managed to fix the issue by disabling all WSL related features, rebooting, then enabling them again and rebooting once more. Later I had to run `wsl --set-default-version 2` again.
> 
> Thank you.... it worked for me..
> 
> 1. wsl --set-default-version 2
> 2. reboot

previous i use the above solution....

but now, the wsl hang issue happened again on my laptop.....and i tried above solution, seems didn't work for me....

i try to find another solution, and found out a new solution that worked for me, as below:

1. Disable WSL windows feature
2. Restart OS
3. Enable WSL windows feature
4. Restart OS

It is going to work again without deleting any of your WSL environments

<img width="1400" height="903" alt="Image" src="https://github.com/user-attachments/assets/c52a9b80-6307-4340-8864-e374f0bf16ad" />

<img width="942" height="791" alt="Image" src="https://github.com/user-attachments/assets/3f85ae1f-91e8-406a-855c-6900870f2fdb" />

**Comment by JuanFontes** [comment]
All commands weren’t working for me, they were all hanging. I restarted, and when I opened VS Code it connected to WSL with no issues. All settings (Hyper-V, Windows Subsystem for Linux, etc.) are correct. Anyway, a restart “fixed” it.

**Comment by paulangton** [comment]
unsubscribe
